proposals/js-string-builtins/Overview.md

```diff
@@ -407,3 +384,54 @@ function compare(
+
+One way we could do this, is by having the JS-API bind the first imported memory of a module to any imported builtin functions that want to operate on memory. If there is no imported memory and a builtin function that needs memory is imported, then a link error is reported.
+
+The memory is imported as opposed to exported so that it is guaranteed to exist when the builtin imports are provided. Using a memory defined only locally would have limited flexibility and would also be exposing a potentially private memory to outside its module.
+
+A quick example:
+```
+(module
+  (; memory 0 ;)
+  (import ... (memory ...))
+
+  (; bound to memory 0 through the JS-API instantiating the builtin module ;)
+  (import "wasm:js-string" "encodeStringToMemoryUTF16" (func ...))
+)
+```
+
+Because the `wasm:js-string` module is instantiated when the module using it is instantiated, the imported memory will be around to be provided to both modules.
+
+If multi-memory is in use and the desired memory to bind with is not the first import, then we could consider parsing the imports to determine which memory is needed for which builtin. For example, `encodeStringToMemoryUTF16.2` for binding to memory 2.
+
+### Better function types to avoid runtime checks
+
+The initial set of JS String Builtins are typed to use `externref` for wherever a JS String is needed. This can lead to runtime checks that should be avoidable. Optimizing compilers can probably get rid of some of these, but not all.
+
+In the future, we could have type imports or a core stringref type. In this event, it would be desirable to use those in the function types to avoid unnecessary runtime checks.
+
+The difficulty is how to do this in a backwards compatible way. If we, for example, changed the type of a builtin from `[externref] -> []` to `(ref null 0) -> []`, we would break old code that imported it with the externref parameter.
+
+One option would be to version the name of the function builtins, and add a new one for the more advanced type signature.
+
+Another option to do this would be to extend the JS-API to inspect the function types used when importing these builtins to determine whether to provide it the 'advanced type' version or the 'basic type' version. This would be a heuristic, something like checking if the type refers to a type import or not.
+
+### UTF-8 and WTF-8 support
+
+There are no JS builtins available to get a UTF-8 or WTF-8 view of a JS String.
+
+One option would be to specify wasm builtins in term of the Web TextEncoder and TextDecoder interfaces. But this is probably a 'layering' violation, and is not clear what this means on JS runtimes outside the web.
+
+Another option around this would be to directly refer to the UTF-8/WTF-8 specs in the JS-API and write out the algorithms we need. However, this probably violates the goal of not creating a new String API.
```

Comment:
I think this option and the one before (line 431) are effectively the same, which solves both listed problems: there's no material difference between saying "do what TextEncoder does" and "perform a UTF-8 encoding", either way it's clear what that means, and it doesn't create a new String API on the web because TextEncoder already does the same thing.

That said, TextEncoder/TextDecoder don't deal with WTF-8, so it'll be a slightly different beast anyway; which once again goes to show that the Strings API that Wasm needs is perhaps somewhat similar to but not the same as the Strings API that JS provides; but you probably don't want to have that discussion here ☺

Reply:
I think where the options concretely differ is that the second option would add a dependency to a web standard in non-web JS environments like Node. And so any changes to the TextEncoder standard would be meaningful in those environments when they may not have been before.

Possibly not a big deal, I just haven't thought much about it or looked into this much.